src/debugger/include/debugger/debugger.h

```diff
@@ -29,47 +30,106 @@ namespace debugger {
 	 */
-	CPU *cpu;
+	std::vector<Address> breakpoints;
 
 	/**
-	 * Pointer to memory instance inside gameboy
+	 * A bool value to specify whether the debugger has currently encountered a
+	 * breakpoint condition and waiting for the user input. This will reset to
+	 * false once run is called again after a breakpoint.
 	 */
-	Memory *memory;
+	bool is_breaking = false;
 
 	/**
-	 * Pointer to GPU instance inside gameboy
+	 * A vector for storing all the processed breakpoints
 	 */
-	GPU *gpu;
+	std::vector<Address> processed_breakpoints;
 
 	/**
-	 * Pointer to Cartridge instance inside gameboy
+	 * A vector container for storing breakpoints at CPU cycles
 	 */
-	Cartridge *cartridge;
+	std::vector<ClockCycles> cycle_breakpoints;
+
+	/**
+	 * A vector container for storing breakpoints at GameBoy cycles
+	 */
+	std::vector<ClockCycles> tick_breakpoints;
+
+	/**
+	 * A vector container iterator to keep track of breakpoints
+	 */
+	std::vector<Address>::iterator breakpoints_iter = breakpoints.begin();
+
+	/**
+	 * A flag to check if we have reached the breakpoint
+	 */
+	bool is_breakpoint = false;
 
   public:
 	/**
-	 * Debugger constructor
+	 * DebuggerCore constructor
 	 */
-	Debugger(std::unique_ptr<gameboy::Gameboy> gameboy);
+	DebuggerCore(std::unique_ptr<gameboy::Gameboy> gameboy);
 
 	/**
 	 * Run debugger iteration
 	 */
-	void tick();
+	virtual void tick();
+
+	/**
+	 * @brief Adds breakpoints to the Vector Container, skips if breakpoint
+	 * already present
+	 * @param breakpoint of Instructions, ticks and clocks
+	 *
+	 * returns 0 if Breakpoint was already present, so not added;
+	 * returns 1 if Breakpoint was successfully inserted;
+	 * returns 2 default return value. Will not reach here;
+	 * I chose uint8_t as a return type so that we can add to the error codes
+	 * here if the need arises in the future.
+	 */
+	virtual uint8_t set_breakpoint(Address breakpoint);
+	virtual uint8_t set_tick_breakpoint(ClockCycles tick);
+	virtual uint8_t set_cycle_breakpoint(ClockCycles cycle);
+
+	/**
+	 * @brief Removes breakpoints from Vector container and returns
+	 * corresponding code
+	 * @param breakpoint of Instructions, ticks and clocks
+	 *
+	 * returns 0 if Breakpoint was not present, so not removed;
+	 * returns 1 if Breakpoint was successfully removed;
+	 * returns 2 default return value. Will not reach here;
+	 * I chose uint8_t as a return type so that we can add to the error codes
+	 * here if the need arises in the future.
+	 */
+	virtual uint8_t remove_breakpoint(Address breakpoint);
+	virtual uint8_t remove_tick_breakpoint(ClockCycles tick);
+	virtual uint8_t remove_cycle_breakpoint(ClockCycles cycle);
+
+	// Get current debugger state
+	virtual vector<Address> get_breakpoints();
+	virtual vector<ClockCycles> get_cycle_breakpoints();
+	virtual vector<ClockCycles> get_tick_breakpoints();
+
+	virtual void run();
+	virtual void step();
+
+	virtual std::string peek(uint32_t lines);
```

Comment:
Perhaps return a std::vector of strings? The debugger core shouldn't do any formatting. That should be done by the UI layer.

You can move the newline printing logic to the CliDebugger class instead.

Reply:
Agreed.

Reply:
I rather use `map` because having a instruction address corresponding to the instruction mnemonic, it helps a lot while setting breakpoints. What do you think?

Reply:
Yup, that's cool